### ISSUE TITLE
QR codes too small

### ISSUE THREAD

**Issue by michael-forsyth** [cta-wave/Test-Content#28]
Two issues been found related to the size of qr codes in the test videos:

 - In the low resolution videos the number of video pixels for each qr pixel with the lowest resolutions having a 1-1 mapping
     - This makes it easy for the qr codes to become corrupted by the encoding process, resizing the video for display or the capture method. avc_sets/15_30_60/t34 is an example that has been has corrupted qr codes from just the encoding process.
    - Pass investigation by Eurofins has found a minimum of 3 video pixels to 1 qr pixel is needed when encoding video that is expected to be captured by a camera.

 - When testing low to mid-range mobile devices it has been found that the low effective display resolution (less than 720p) combine with small screen size and non-integer mapping of display to qr pixels, significantly degrades the qr_codes to the point where detection is often not practical for higher resolutions or possible for the lower resolutions.

The simplest solution to both issues appears to be increasing the size of the qr codes in the video.

**Comment by jpiesing** [comment]
@nicholas-fr?

**Comment by jpiesing** [comment]
It may be that we have to drop the lowest resolutions as I don't think there is screen real-estate available to double or triple the size of the QR codes.

**Comment by nicholas-fr** [comment]
Indeed, looking at the space used for the QR codes and the adjacent annotations in the lowest resolutions (e.g. 480x270), we can't make any meaningful increase in QR code size to address this. 

One alternative would be to alter the error correction level to reduce the QR code pixels from 33x33 to 29x29 or 25x25, but then the error correction is less robust. However, that would first require further investigation, to ensure it does not negatively impact the detection reliability in the observation framework.

Looking at [screen sizes and pixel densities](https://developer.android.com/about/dashboards) of active Android devices in the Google Play ecosystem, such lower-end smartphones (<720p, low pixel density) are no longer common. So I wouldn't propose making significant changes for the lowest resolutions on those devices.

So in summary: simply skipping the lowest resolution(s) seems the simplest solution at this point.

**Comment by gitwjr** [comment]
@jpiesing Can we close this issue? Seems at this point we have defaulted to skipping lowest res rather than investigating changes to the QR codes. What would that impact? Seems less content to gen/regen/validate/verify; perhaps fewer tests to run; changes to the Spec? Or has this already been done?

**Comment by jpiesing** [comment]
It's not my issue but as far as I'm concerned, yes we can close it.

**Comment by gitwjr** [comment]
@michael-forsyth @louay @nicholas-fr 
Assuming we can close this issue, are there any (positive) impacts for ongoing progress toward Launch as I mentioned above? i.e. reduction of things that need to be done/completed?

**Comment by nicholas-fr** [comment]
I propose to check with Eurofins. @yanj-github Do you have any concerns or do you have more info on detection issues related to the lower resolution test streams?

To decide whether or not to drop one or more of the lowest resolutions, I believe more testing feedback about this problem would be needed. 

Dropping the lowest resolution(s) would also require creating at least one new test stream, for the lowest frame rate (e.g. 12.5 fps, 15fps).
So removing streams would create more work (i.e. replacement for t34 test steam with a higher res, and update of ss1 switching set).

**Comment by gitwjr** [comment]
@nicholas-fr @yanj-github 
I was only looking for opportunities to reduce the work, not increase it. It seems I misunderstood Jon's comment above. 
"It may be that we have to drop the lowest resolutions as I don't think there is screen real-estate available to double or triple the size of the QR codes.."

What, if anything, needs to be done to address this issue and close it?

**Comment by yanj-github** [comment]
@nicholas-fr  and @gitwjr We dont think lower resolution test streams can be reliabily tested.

**Comment by jpiesing** [comment]
> @nicholas-fr and @gitwjr We dont think lower resolution test streams can be reliabily tested.

Reading the comments above, I thought this only applied to "low to mid-range mobile devices" in which case we could just decide to ignore those devices.

**Comment by nicholas-fr** [comment]
That was my understanding. Unfortunately we didn't receive any feedback from the author of the issue.

To have actionable feedback I believe we need to know:
- Which "lower resolution test streams".
- Whether this is only a problem under specific conditions, e.g. only on certain "low to mid-range mobile devices" or not.
- What testing data is available to confirm this is a problem.

**Comment by yanj-github** [comment]
This unfortunately applies to all the DUT. What we were trying to say is that qr codes might became degraded or corrupted by the encoding process. Mike F. checked it before, but that time was with old version of content. And I have run similar test again with the latest content. We used ffmpeg to get png image out from video content and use pyzbar (same way we detected qr code on OF) to check if qr code are reliably and accurately detected.
I have tested on t34 stream from this location https://dash.akamaized.net/WAVE/vectors/avc_sets/15_30_60/t34/.  It was unable to detect all the QR codes.
We didn’t check others, but I guess other streams encoded with low resolution A1 mezzanine probably have the same issue.

**Comment by nicholas-fr** [comment]
Decision Telco 2022-10-19 : 
This will not hold back any release. Problem content will be removed and we will seek to address the issue in a future update.

@yanj-github Would Eurofins be able to test the next lowest resolutions as well? Hopefully only 1 or 2 additional resolutions need testing.
You've tested t34, so next would be t32, then t31, t30 and so on. As soon as robust detection is possible, that would be the lowest resolution to include.

**Comment by yanj-github** [comment]
I have checked only **15_30_60** group:

- t34: 826 frames detected / total 900 frames is expected
- t32: 1580 frames detected / total 1800 frames is ecpected
- t31: all frames detected
- t30: all frames detected

I think anything that uses A1 has a problem, but I didnt test them all.

**Comment by gitwjr** [comment]
A1 should be eliminated for the launch. Fix in next release. 
Determine if any other framerate groups use B1 for the t32 vector.

**Comment by yanj-github** [comment]
Sorry for the confision, now I can confirm t32 uses B1 for 15_30_60 group for 2022-10-17 release. 

I have retest this by clean download from the latest release which is https://dash.akamaized.net/WAVE/vectors/cfhd_sets/15_30_60/t32/2022-10-17/

-  **t34: 826 frames detected / total 900 frames is expected (A1)**
-  **t32: 1774 frames detected / total 1800 frames is ecpected (B1)**
-  t31: all frames detected (C1)
-  t30: all frames detected (D1)

So contents from both A1 and B1 are having issues.

**Comment by gitwjr** [comment]
@yanj-github @nicholas-fr @jpiesing 
"Decision Telco 2022-10-19 :
This will not hold back any release. Problem content will be removed and we will seek to address the issue in a future update."
1) Does this mean we can close this issue? 
2) Do we need to document that certain resolutions will not be included in certain tests? If so, who will document and where?

**Comment by gitwjr** [comment]
Closed. Reopen if necessary.
